Created pushHeadless sample app for push registration via headless on INSTALL vs on first launch.

jamesjhedley opened the pull request.

kwallis commented:
Does it only run on install? It would need to run every time the device boots as well would it not?

jamesjhedley commented:
Well technically the registration should only ever need to run once (on install) but yea the headless app should also run after each device boot :).

Commit:
Created pushHeadless sample app for push registration via headless on INSTALL vs on first launch.
-Includes WebWorks portion to receive and view pushes
-Utilizes config.ini file for push credentials to avoid having to
rebuild native component
-Includes README for setup and information on how to send pushes for BES and BIS

bryanhiggins commented on pushHeadless/README.md:
```diff
@@ -0,0 +1,86 @@
+# Push Headless Sample Application
+
+The Push Headless sample push-enabled application demonstrates how to write a BlackBerry 10 WebWorks application that uses a backend headless service to register for push. This allows registration to be done on app install instead of on-first-run. The webworks portion will receive any pushes.
```
Capitalization of WebWorks is incorrect

bryanhiggins commented on pushHeadless/README.md:
```diff
@@ -0,0 +1,86 @@
+The Push Headless sample push-enabled application demonstrates how to write a BlackBerry 10 WebWorks application that uses a backend headless service to register for push. This allows registration to be done on app install instead of on-first-run. The webworks portion will receive any pushes.
+
+Consult the [PushService](http://developer.blackberry.com/html5/apis/beta/blackberry.push.pushservice.html) and [PushPayload](http://developer.blackberry.com/html5/apis/beta/blackberry.push.pushpayload.html) classes from the API reference for useful examples and a full description of how to use the
+push APIs.
```
Why is there a line break here?

bryanhiggins commented on pushHeadless/README.md:
```diff
@@ -0,0 +1,86 @@
+
+1. An overview of push and the Push Service architecture
+2. The requirements for creating a full push solution
+3. How to download, build, and configure the sample application when it's loaded on your BlackBerry 10 device (This is also described below.)
```
remove comma after build

bryanhiggins commented on pushHeadless/pushCollectorHeadless/bar-descriptor.xml:
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="utf-8" standalone="no"?>
+<!--
+  
+   Copyright (c) 2011, 2012, 2013 Research In Motion Limited.
```
2014 BlackBerry Ltd

bryanhiggins commented on pushHeadless/pushCollectorHeadless/bar-descriptor.xml:
```diff
@@ -0,0 +1,99 @@
+    <permission system="true">run_native</permission>
+    <env var="LD_LIBRARY_PATH" value="app/native/lib:/usr/lib/qt4/lib"/>
+    
+</qnx>
```
is pushCollectorHeadless/icon.png used by WebWorks?

jamesjhedley replied:
This file will end up being removed. It's not needed.

bryanhiggins commented on pushHeadless/pushCollectorHeadless/readme.txt:
```diff
@@ -0,0 +1,33 @@
+including:
+
+- Creating and destroying channels
+- Processing incoming push notifications
```
Isn't the point of this that WebWorks processes incoming push and the native side only does registration?

jamesjhedley replied:
Missed this. Should be removed or updated to reflect our sample.

bryanhiggins commented on pushHeadless/pushCollectorHeadless/readme.txt:
```diff
@@ -0,0 +1,33 @@
+Running the example:
+
+The headless application cannot be launched directly by the user. To
+launch it, run the Push Collector application. The Push Collector uses
```
Is this relevant to WebWorks?

bryanhiggins commented on pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp:
```diff
@@ -0,0 +1,264 @@
+    // If doing a BIS push, comment out the initConfigForBES call in favour of the initConfigForBIS call.
+    // Don't forget to verify your settings in the below two calls.
+    initConfigForBES();
+    //	initConfigForBIS();
```
Can this not be done by configuration? This implies that in order to use consumer push, the WebWorks dev needs to download Momentics, copy the project over, import it, uncomment this line and compile.

jamesjhedley replied:
Technically they just need to update the config elements in this file (or a property file) and run make in the pushHeadless directory. Make comes with bbndk i believe. They however, may not have bbndk.  I'll definitely look into possibly using some sort of property file that can be read at runtime to avoid any rebuilding of the native portion.

bryanhiggins commented on pushHeadless/pushCollectorHeadless/src/AppHeadless.cpp:
```diff
@@ -0,0 +1,264 @@
+
+void AppHeadless::initConfigForBES()
+{
+    m_configuration->setProviderApplicationId("myAppID");
```
All of these hard coded values should be part of configuration.

jamesjhedley replied:
The reason i left it separate was to allow the developer to clearly see what is/isn't required for each type of push. When i was learning push i was VERY confused as to what i actually needed to provide when configuring. For BES all you really need is an AppId, unless you are working with the Push SDK and have an initiator URL and credentials. I just don't want to give them a configuration page with a bunch of empty properties and hope they know what to fill in for each.

I do however agree that we should try and make this as easy for the user as possible. No building, Easy setup.
In order to do that we need to distinguish what fields are required for each type of push and when each optional parameter is needed.

bryanhiggins commented on pushHeadless/www/config.xml:
```diff
@@ -0,0 +1,53 @@
+                <entryPointType>Qnx/Elf</entryPointType>
+                <permission system="true">_sys_run_headless</permission>
+                <permission>read_device_identifying_information</permission>
+                <permission>access_shared</permission>
```
What does it need access_shared for?

jamesjhedley replied:
Shouldnt need this. Good catch

bryanhiggins commented on pushHeadless/www/config.xml:
```diff
@@ -0,0 +1,53 @@
+            </invoke-target>
+        </config-file>
+    </platform>
+</widget>
```
Shouldn't use_consumer_push be somewhere in this file along with instructions to uncomment for enterprise push?

jamesjhedley replied:
Definitely a good idea.

bryanhiggins commented on pushHeadless/www/index.html:
```diff
@@ -0,0 +1,44 @@
+        <title>Hello World</title>
+    </head>
+    <body>
+        <div class="app">
```
This could be cleaned up. I don't think its necessary to show the Apache Cordova logo or "WebWorks Ready".

jamesjhedley replied:
Hmm, i'm torn on this one. If they are new to WebWorks i can see this definitely being useful and If they are a WebWorks Vet they are going to simply replace their www with ours anyways. But, yea; None of the other samples keep the logo and webworks ready stuff. Probably is a good idea to keep it as clean as possible and remove.

bryanhiggins commented on pushHeadless/www/js/index.js:
```diff
@@ -0,0 +1,97 @@
+        var ops = {
+            invokeTargetId : "com.example.pushCollector.invoke.push",
+            appId : "myAppID"
+        };
```
It would be better to read this from configuration.

jamesjhedley replied:
Not sure i fully understand what you mean. Do you mean it would be better to read this from the proposed "configuration property file"? Would be awesome if we could set it up so both the web and native code read from the same file. One stop configuration. No editing several files like all the other samples.

bryanhiggins commented:
The native side should be reviewed by Shadid. Sorry, I don't know his GitHub user name.

jamesjhedley commented:
Still need to update WebWorks side to also read from config.

jamesjhedley commented:
Before merging this needs to have the changes from here ported https://github.com/blackberry/webworks-cli/pull/18/files

jamesjhedley commented:
BIS push passed for me. Getting steve to test as well.

jamesjhedley commented:
Rebased, Squashed.
Can someone take a look at the README and see if i missed anything? After that we should be good to merge.

jamesjhedley commented:
Can somebody please merge this and get it into blackberry repo?
